### DISCUSSION TITLE
How do I rewrite in Build Output API edge middleware??

### DISCUSSION THREAD

**Discussion by AaronBeaudoin** [vercel/vercel#8423] [Help] [answered]
**We are not talking about `NextResponse.rewrite` here. We're talking about the Build Output API directly.**

I have the following important files:

```js
// .vercel/output/config.json

{
  "version": 3,
  "routes": [
    { "src": "/fn/(.*)", "middlewarePath": "_middleware/index", "continue": true },
    { "handle": "filesystem" },
    { "src": "/(.*)", "dest": "/_pages" }
  ]
}
```

```js
// .vercel/output/functions/_middleware/index.func/index.js

export default function middleware(request) {
  const url = new URL(request.url);
  url.pathname = "/";

  const response = new Response();
  // response.headers.set("x-middleware-next", "1");
  response.headers.set("x-middleware-rewrite", url.toString());
  response.headers.set("x-spaghetti", "yes");
  return response;
}
```

1. `/_pages` is an edge function that renders my pages.
2. `x-spaghetti` is just there for testing.
3. Going to `/` gives me my homepage.
4. I expect going to `/fn/anything` should also give me my homepage?
5. Actually, going to `/fn/anything` gives me a 404, indicating that my rewrite isn't working.
6. I know the middleware is running because I get `x-spaghetti` in my browser.
7.  **What am I missing here???**

I tried:

- Moving around the order of my routes in my config.
- Toggling `x-middleware-next`.
- Toggling `"continue": true` in my route.
- Other stuff. I'm out of ideas.

**Edit 1:**

I've also tried setting up [`vercel/examples/edge-functions/ab-testing-simple/`](https://github.com/vercel/examples/tree/main/edge-functions/ab-testing-simple) and then running `vercel build` to get the `.vercel/` directory and see the Build Output API files that get created. So far it's been totally unhelpful. It really seems like my project should work. The only thing I've found in the example that I don't understand is that the route for the middleware in `config.json` gets an `"override": true` property which as far as I can tell is 100% undocumented anywhere. Thanks Vercel!

**Edit 2:**

Also, for the [`vercel/examples/build-output-api/edge-middleware`](https://github.com/vercel/examples/tree/main/build-output-api/edge-middleware) example, when I go to https://build-output-api-edge-middleware.vercel.sh I totally am not seeing a `x-modified-edge` header in my network tab in devtools... **is edge middleware just broken or something???** It kinda seems like there's an invisible `{ "handle": "filesystem" }` maybe taking precedence over the middleware in that example?

**Comment by amyegan** [comment]
Hi @AaronBeaudoin! Can you share the output of your Project from the settings page? You can find it by navigating to your Project page -> Deployments tab -> Deployment details -> Output tab. That will tell is if the middleware is being recognized in your Project. 

For example, here's what I have in my test app (was not able to reproduce the problem from the example repo)
<img width="797" alt="image" src="https://user-images.githubusercontent.com/8247105/185655432-8541f2f5-4539-4fa0-bda0-3a7aae9c0f5d.png">


Regarding `Edit 2`, I am definitely seeing `x-modified-edge` in response headers for https://build-output-api-edge-middleware.vercel.sh/

<img width="332" alt="image" src="https://user-images.githubusercontent.com/8247105/185635435-faeb9427-749f-49e0-a37a-22e3632c59cc.png">

I'm wondering if maybe there is some kind of difference in our networks or computer configurations? Please share a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) (or your repo if it's public). That way we can take a closer look and see exactly what is going wrong for your app.

**Reply by AaronBeaudoin** [reply]
Hi @amyegan. Here is my minimal example: https://github.com/AaronBeaudoin/vercel-edge-middleware-test

See my comment below for more information.

**Edit:** I've rewritten this example quite a bit since I wrote this, and created a new [more specific question](https://github.com/vercel/vercel/discussions/8487) that is more focused on exactly what I'm trying to accomplish here.

**Reply by AaronBeaudoin** [reply]
Also, regarding **Edit 2** it looks like when I do <kbd>Command</kbd>+<kbd>Shift</kbd>+<kbd>R</kbd> to hard refresh the page I can see `x-vercel-cache` in my devtools, but when I just do <kbd>Command</kbd>+<kbd>R</kbd> to refresh regularly it is not there. It looks like this is because Vercel adds an `etag` header which causes my browser to send a `if-none-match` header the next time I make a request to the page. The server then responds with a `304` response and the `x-vercel-cache` is not included. Can you verify this?

**Reply by AaronBeaudoin** [reply]
Alright, I think I've got a clearer picture of exactly what the limitation I'm facing here is.

Feel free to look at the minimal example to see what I'm working with, but be aware it may change as I continue messing around.

## The Setup

Suppose you've got 3 edge functions:

1. `_default`: A render function to be used to render pages by default.
2. `_alternate`: An alternate render function for some specific pages.
3. `_middleware`: An edge function to be used as middleware before calling `_alternate`.

In `_middleware` you have `response.headers.set("x-middleware-rewrite", "/_alternate")` in order to call the `_alternate` edge function afterwards.

And you have the following `.vercel/output/config.json`:

```json
{
  "version": 3,
  "routes": [
    { "src": "/test", "middlewarePath": "_middleware" },
    { "src": "/(.*)", "dest": "/_default" }
  ]
}
```

The idea here is that all pages except `/test` should go to `_default`. The `/test` page should go to `_alternate`, but the rewrite should be handled inside of the `_middleware` function.

## The Problem

When `_alternate` runs via a request to `/test` through the `_middleware` function, `request.url` is still `http://my-site.com/test`. What I would like to do is modify the URL/path received by the `_alternate` function. So far I have tried two things, both of which didn't work:

1. Setting the `Location` header inside `_middleware`. It had no effect.
2. Setting some arbitrary `x-custom-url` header inside `_middleware` and then trying to read it from `_alternate`. This didn't work because no `x-custom-url` header appears to be undefined when `_alternate` runs, so it seems that the header is not finally set until after the final edge function runs, which seems weird. Not sure if I'm missing some important detail here.

## The Goal

Via an edge middleware function, I'd like to no only rewrite to another URL (edge function) in my Vercel project, but also rewrite the URL that will be seen by that function (or somehow pass a new URL to the function). I feel like this should be possible, so I'm not sure what I'm missing here. Any help would be much appreciated.

**Comment by AaronBeaudoin** [comment, marked as answer]
_The problem I'm facing is actually a bit different from what I thought when writing this question. See https://github.com/vercel/vercel/discussions/8487 for a new more specific question that better expresses exactly the issue I'm seeing._

If you're having the "same problem" as this question, however, hopefully this helps: The `x-middleware-rewrite` header used in my question above is actually working "as intended", but the 404 occurs because of the following:

- The middleware sets `x-middleware-rewrite` to rewrite back to the home page at `/`.
- The `_pages` edge function runs as a result, but `request.url` is not `/` as I had expected, but still `/fn/anything`, so naturally the render function chokes and throws a 404 because it doesn't have a matching page for the route `/fn/anything`.

The key takeaway is that rewriting via the `x-middleware-rewrite` header only sets **what resource on the filesystem Vercel should "rewrite" to**. It does not actually change `request.url` for that "next" resource. The new question (https://github.com/vercel/vercel/discussions/8487) mentioned above is about how I can actually pass on some arbitrary data (such as a new URL) to the resource that will be run.
